\documentclass[11pt, oneside]{article}
\usepackage{geometry}
\geometry{letterpaper}
\usepackage{graphicx}
\usepackage{amssymb}
\usepackage[affil-it]{authblk}

\title{CompGO vignette}
\author{Sam Bassett, Ash Waardenberg}
\affil{Developmental and Stem Cell Biology Lab,\\Victor Chang Cardiac Research Institute,\\Darlinghurst, Sydney, Australia}
\date{}

\begin{document}
\maketitle
%\VignetteIndexEntry{Introduction}
\section{Introduction}
This package contains functions to accomplish several tasks relating to gene ontology enrichment comparison and visualisation. It interfaces with rtracklayer and VariantAnnotation to easily annotate .bed files with genes, and with RDAVIDWebService to generate functional annotation charts based on these gene lists. From here, we provide several methods for visualisation, including viewing the GO term hierarchy in two samples using a directed acyclic graph (DAG), performing a z-score standardisation on the GO terms returned from DAVID, comparing these directly using a pairwise scatterplot with linear fit and Jaccard metric, performing a Kolomogorov-Smirnov test against the distributions to determine their similarity, and clustering multiple inputs into a dendrogram.

%\VignetteIndexEntry{Dumping UCSC databases}
\section{Dumping UCSC databases}
We provide the UCSC mm9 refGene database as a data file, but if a different track is needed this function can be used to do so. It can sometimes take some time to run, since it's downloading a large file.

<<>>=
## Get UCSC database for human genome, refGene track
library(CompGO)
#db = ucscDbDump(genome="hg19") # instead of running this
    ## unnecessarily, the next statement is equivalent:
data(ucsc.mm9)
db = ucsc.mm9
str(db)
@

%\VignetteIndexEntry{.bed file importing and annotation}
\section{.bed file importing and annotation}
Reading a .bed file into R is easy:
\begin{verbatim}
## Not run, needs path
bed = read.bed('./path_to_file.bed')
## Sometimes, such as when using biomaRt, the chromosome
## name must be removed, so an option exists to do so
bed = read.bed('./path_to_file.bed', subChr=TRUE)
\end{verbatim}
Once a .bed file is read, you can start annotation. Here we use the sample .bed file provided.
<<>>=
data(bed.sample)
## This step can take some time
geneList = annotateBedFromUCSC(bedfile=bed.sample,db=ucsc.mm9)
str(geneList)
@

%\VignetteIndexEntry{Functional Enrichment and visualisation}
\section{Functional Enrichment and visualisation}
This gene list can then be passed on to DAVID for gene ontology enrichment analysis. This returns an object of type DAVIDFunctionalAnnotationChart, but it can be accessed similarly to a data.frame.
\begin{verbatim}
## Not run, since it needs a registered email
fnAnot = getFnAnot_genome(geneList$name2, email="your_registered@email.com",
+	idType="REFSEQ_MRNA",listName="sample1")
str(fnAnot)
\end{verbatim}

After obtaining an enrichment table, we can visualise differences between two sets of genes (or two .bed files) using either a pairwise scatterplot or a combination directed acyclic graph (DAG), where each node (representing an individual GO term) is coloured based on which set it came from.
\begin{verbatim}
## Since this requires email registration, the following won't be run.
anot1 = getFnAnot_genome(geneList1, email="email@email.com")
anot2 = getFnAnot_genome(geneList2, email="email@email.com")

## To produce scatterplot:
plotPairwise(anot1, anot2, cutoff = 0.05) # using benjamini as metric
# plotting all GO terms, even if they only appear in one set
plotPairwise(anot1, anot2, plotNA=TRUE, cutoff=0.1)

## For DAGs:
plotTwoGODags(anot1, anot2) # no relaxation, default colours
plotTwoGODags(anot1, anot2, relaxPvals = TRUE) # with relaxation
\end{verbatim}


\end{document}
